core/src/main/java/org/infinispan/persistence/support/NonBlockingStoreAdapter.java

```diff
@@ -0,0 +1,340 @@
+package org.infinispan.persistence.support;
+
+import java.lang.invoke.MethodHandles;
+import java.util.EnumSet;
+import java.util.Set;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import javax.transaction.Transaction;
+
+import org.infinispan.commons.api.Lifecycle;
+import org.infinispan.commons.dataconversion.MediaType;
+import org.infinispan.commons.persistence.Store;
+import org.infinispan.commons.util.IntSet;
+import org.infinispan.persistence.spi.AdvancedCacheExpirationWriter;
+import org.infinispan.persistence.spi.AdvancedCacheLoader;
+import org.infinispan.persistence.spi.AdvancedCacheWriter;
+import org.infinispan.persistence.spi.CacheLoader;
+import org.infinispan.persistence.spi.CacheWriter;
+import org.infinispan.persistence.spi.FlagAffectedStore;
+import org.infinispan.persistence.spi.InitializationContext;
+import org.infinispan.persistence.spi.MarshallableEntry;
+import org.infinispan.persistence.spi.MarshallableEntryFactory;
+import org.infinispan.persistence.spi.NonBlockingStore;
+import org.infinispan.persistence.spi.SegmentedAdvancedLoadWriteStore;
+import org.infinispan.persistence.spi.TransactionalCacheWriter;
+import org.infinispan.reactive.RxJavaInterop;
+import org.infinispan.util.concurrent.BlockingManager;
+import org.infinispan.util.concurrent.CompletableFutures;
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+import org.reactivestreams.Publisher;
+
+import io.reactivex.rxjava3.core.Flowable;
+import io.reactivex.rxjava3.processors.FlowableProcessor;
+import io.reactivex.rxjava3.processors.UnicastProcessor;
+
+public class NonBlockingStoreAdapter<K, V> implements NonBlockingStore<K, V> {
+   private static final Log log = LogFactory.getLog(MethodHandles.lookup().lookupClass());
+   private static final boolean trace = log.isTraceEnabled();
+   private static final AtomicInteger id = trace ? new AtomicInteger() : null;
+
+   private final Lifecycle oldStoreImpl;
+   private final Set<Characteristic> characteristics;
+
+   private BlockingManager blockingManager;
+   private MarshallableEntryFactory<K, V> marshallableEntryFactory;
+
+   public NonBlockingStoreAdapter(Lifecycle oldStoreImpl) {
+      this.oldStoreImpl = oldStoreImpl;
+      this.characteristics = determineCharacteristics(oldStoreImpl);
+   }
+
+   public Lifecycle getActualStore() {
+      return oldStoreImpl;
+   }
+
+   private String nextTraceId() {
+      return id != null ? "StoreAdapter-" + id.getAndIncrement() : null;
+   }
+
+   static private Set<Characteristic> determineCharacteristics(Object storeImpl) {
```

Comment:
:vomiting_face:  ... shows how crazy it was before!

Reply:
Yeah, blargh.